cloudnew/ec2.go

```diff
@@ -516,3 +518,170 @@ func (m *ec2Manager) getSpotInstanceStatus(id string) (cloud.CloudStatus, error)
+	os := osLinux
+	if strings.Contains(h.Distro.Arch, "windows") {
+		os = osWindows
+	}
+
+	dur := end.Sub(start)
+	region := azToRegion(*instance.Placement.AvailabilityZone)
+	iType := instance.InstanceType
+
+	ebsCost, err := m.blockDeviceCosts(instance.BlockDeviceMappings, dur)
+	if err != nil {
+		return 0, errors.Wrap(err, "calculating block device costs")
+	}
+	hostCost, err := onDemandCost(&pkgOnDemandPriceFetcher, os, *iType, region, dur)
+	if err != nil {
+		return 0, errors.WithStack(err)
+	}
+	return hostCost + ebsCost, nil
+}
+
+func (m *ec2Manager) costForDurationSpot(h *host.Host, start, end time.Time) (float64, error) {
+	spotDetails, err := m.client.DescribeSpotInstanceRequests(&ec2.DescribeSpotInstanceRequestsInput{
+		SpotInstanceRequestIds: []*string{makeStringPtr(h.Id)},
+	})
+	if err != nil {
+		return 0, err
+	}
+	instance, err := m.getInstanceInfo(*spotDetails.SpotInstanceRequests[0].InstanceId)
+	if err != nil {
+		return 0, err
+	}
+	os := osLinux
+	if strings.Contains(h.Distro.Arch, "windows") {
+		os = osWindows
+	}
+	ebsCost, err := m.blockDeviceCosts(instance.BlockDeviceMappings, end.Sub(start))
+	if err != nil {
+		return 0, errors.Wrap(err, "calculating block device costs")
+	}
+	spotCost, err := m.calculateSpotCost(instance, os, start, end)
+	if err != nil {
+		return 0, err
+	}
+	return spotCost + ebsCost, nil
+}
+
+func (m *ec2Manager) blockDeviceCosts(devices []*ec2.InstanceBlockDeviceMapping, dur time.Duration) (float64, error) {
+	cost := 0.0
+	if len(devices) > 0 {
+		volumeIds := []*string{}
+		for i := range devices {
+			volumeIds = append(volumeIds, devices[i].Ebs.VolumeId)
+		}
+		vols, err := m.client.DescribeVolumes(&ec2.DescribeVolumesInput{
+			VolumeIds: volumeIds,
+		})
+		if err != nil {
+			return 0, err
+		}
+		for _, v := range vols.Volumes {
+			// an amazon region is just the availability zone minus the final letter
+			region := azToRegion(*v.AvailabilityZone)
+			p, err := ebsCost(&pkgEBSFetcher, region, *v.Size, dur)
+			if err != nil {
+				return 0, errors.Wrapf(err, "EBS volume %v", v.VolumeId)
+			}
+			cost += p
+		}
+	}
+	return cost, nil
+}
+
+// calculateSpotCost is a helper for fetching spot price history and computing the
+// cost of a task across a host's billing cycles.
+func (m *ec2Manager) calculateSpotCost(
+	i *ec2.Instance, os osType, start, end time.Time) (float64, error) {
+	rates, err := m.describeHourlySpotPriceHistory(
+		*i.InstanceType, *i.Placement.AvailabilityZone, os, *i.LaunchTime, end)
+	if err != nil {
+		return 0, err
+	}
+	return spotCostForRange(start, end, rates), nil
+}
+
+// describeHourlySpotPriceHistory talks to Amazon to get spot price history, then
+// simplifies that history into hourly billing rates starting from the supplied
+// start time. Returns a slice of hour-separated spot prices or any errors that occur.
+func (m *ec2Manager) describeHourlySpotPriceHistory(
+	iType string, zone string, os osType, start, end time.Time) ([]spotRate, error) {
```

Comment:
can we have the arguments to this be in some kind of struct, to increase readability?